s3/src/s3_lib.py

```diff
@@ -0,0 +1,372 @@
+        if contains_required_options:
+            getattr(self.on, "s3_connection_info_changed").emit(
+                event.relation, app=event.app, unit=event.unit
+            )
+        else:
+            logger.warning(
+                f"Some mandatory fields: {missing_options} are not present, do not emit credential change event!"
+            )
+
+    def _on_secret_changed_event(self, event: SecretChangedEvent) -> None:
+        """Event handler for handling a new value of a secret."""
+        if not event.secret.label:
+            return
+
+        relation = self.relation_data._relation_from_secret_label(event.secret.label)
+        if not relation:
+            logging.info(
+                f"Received secret {event.secret.label} but couldn't parse, seems irrelevant."
+            )
+            return
+
+        if event.secret.label != self.relation_data._generate_secret_label(
+            relation.name,
+            relation.id,
+            "extra",
+        ):
+            logging.info("Secret is not relevant for us.")
+            return
+
+        if relation.app == self.charm.app:
+            logging.info("Secret changed event ignored for Secret Owner")
+
+        remote_unit = None
+        for unit in relation.units:
+            if unit.app != self.charm.app:
+                remote_unit = unit
+
+        # check if the mandatory options are in the relation data
+        contains_required_options = True
+        credentials = self.get_s3_connection_info()
+        missing_options = []
+        for configuration_option in S3_REQUIRED_OPTIONS:
+            if configuration_option not in credentials:
+                contains_required_options = False
+                missing_options.append(configuration_option)
+
+        # emit credential change event only if all mandatory fields are present
+        if contains_required_options:
+            getattr(self.on, "s3_connection_info_changed").emit(
+                relation, app=relation.app, unit=remote_unit
+            )
+        else:
+            logger.warning(
+                f"Some mandatory fields: {missing_options} are not present, do not emit credential change event!"
+            )
+
+    def _on_relation_broken_event(self, event: RelationBrokenEvent) -> None:
+        """Event handler for handling relation_broken event."""
+        logger.info("S3 relation broken...")
+        getattr(self.on, "s3_connection_info_gone").emit(
+            event.relation, app=event.app, unit=event.unit
+        )
+
+    @property
+    def relations(self) -> list[Relation]:
+        """The list of Relation instances associated with this relation_name."""
+        return list(self.charm.model.relations[self.relation_name])
+
+
+class S3Requires(S3RequirerData, S3RequirerEventHandlers):
+    """The requirer side of S3 relation."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str,
+        container: str | None = None,
```

Comment:
If this code intended to run on 20.04 (py3.8) charm containers, may need to avoid using [PEP604](https://peps.python.org/pep-0604/) (py3.10)?

Reply:
Enforced py38 compatible typing in this file in 416dddf6b80dec139a007dce0561e4a78f3b809c